packages/config/__tests__/Config.test.ts

```diff
@@ -0,0 +1,306 @@
+
+        it("should load a config and populate missing defaults", async () => {
+            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.json");
+            jest.spyOn(fs, "existsSync")
+                .mockReturnValueOnce(false)     // Project user layer
+                .mockReturnValueOnce(true)      // Project layer
+                .mockReturnValueOnce(false)     // User layer
+                .mockReturnValueOnce(false);    // Global layer
+            jest.spyOn(fs, "readFileSync").mockReturnValueOnce("{}");
+            const config = await Config.load(MY_APP);
+            expect(config.properties).toMatchSnapshot();
+            expect(config.properties.defaults).toEqual({});
+            expect(config.properties.profiles).toEqual({});
+            expect(config.properties.plugins).toEqual([]);
+            expect(config.properties.secure).toEqual([]);
+        });
+
+        it("should fail to load config that is not JSON", async () => {
+            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.json");
+            jest.spyOn(fs, "existsSync")
+                .mockReturnValueOnce(false)     // Project user layer
+                .mockReturnValueOnce(true)      // Project layer
+                .mockReturnValueOnce(false)     // User layer
+                .mockReturnValueOnce(false);    // Global layer
+            jest.spyOn(fs, "readFileSync").mockReturnValueOnce("This is not JSON");
+            let error: any;
+            try {
+                await Config.load(MY_APP);
+            } catch (err) {
+                error = err;
+            }
+            expect(error).toBeDefined();
+            expect(error.message).toContain("error reading config file");
+            expect(error.message).toContain(__dirname + "/__resources__");
+            expect(error instanceof ImperativeError).toBe(true);
+        });
+    });
+
+    it("should find config that exists if any layers exist", () => {
+        const config = new (Config as any)();
+        config._layers = [
+            { exists: false },
+            { exists: true },
+            { exists: false }
+        ];
+        expect(config.exists).toBe(true);
+    });
+
+    it("should not find config that exists if no layers exist", () => {
+        const config = new (Config as any)();
+        config._layers = [ { exists: false } ];
+        expect(config.exists).toBe(false);
+    });
+
+    it("should provide a deep copy of layers", () => {
+        const config = new (Config as any)();
+        config._layers = {};
+        config.layers.properties = {};
+        expect(Object.keys(config._layers).length).toBe(0);
+    });
+
+    describe("set", () => {
+        beforeEach(() => {
+            jest.spyOn(Config, "search").mockReturnValue(__dirname + "/__resources__/project.config.user.json");
+            jest.spyOn(fs, "existsSync").mockReturnValueOnce(true).mockReturnValue(false);
```

Comment:
I assume this indicates a project user layer ???

Reply:
Yes, since project user layer is the first one loaded, it's the easiest file to load when mocking `Config`.